### PR TITLE
[DS-1396] Create action logic

### PR DESCRIPTION
**Pull request by ryanwilsonperkin**
This repo seeks to replace https://github.com/waveaccounting/wave-migbot which has similar functionality but is running as a Probot app on Heroku. This will run the same feature set, but within a GitHub action that can be easily installed into any of our applications.

A notable difference from the original is that this one is configured to run either against a Rails or a Django app using the `project-type` input for the action.

## Testing

The following Pull Requests in an example repository demonstrate this check working against both Django and Rails applications:
- https://github.com/waveaccounting/test-migration-check-action/pull/2
- https://github.com/waveaccounting/test-migration-check-action/pull/3
- https://github.com/waveaccounting/test-migration-check-action/pull/4
- https://github.com/waveaccounting/test-migration-check-action/pull/5

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
+const core = require('@actions/core');
```
Comment (ryanwilsonperkin) [inline index.js:1]: Compare this module against the equivalent in https://github.com/waveaccounting/wave-migbot/blob/master/index.js

File: README.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
+# migration-check-action
+
+A GitHub action that blocks Pull Requests that contain changes to both migration files and code. Supports Django and Rails.
+
+## Usage
+
+This check is installed as a GitHub action and will require you to write a workflow file in your repository. Create a file called `.github/workflows/migration-check.yml` with the following content:
```
Comment (eballerini) [inline README.md:7]: If a new file has to be created for each repository in order for this check to run, do you think this is going to be a barrier for adoption?

File: index.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  console.log("Code removed", codeRemoved)
+  console.log("Code modified", codeModified)
+
+  const migrationHasBeenAdded = migrationsAdded.length > 0;
+  const codeHasBeenAdded = codeAdded.length > 0;
+  const codeHasBeenModified = codeModified.length > 0;
+  if (migrationHasBeenAdded && (codeHasBeenAdded || codeHasBeenModified)) {
+    core.setFailed(`Both migrations and code changes detected`);
+  }
+}
```
Comment (eballerini) [inline index.js:90]: What happened to the comment that we leave in case there is a migration?

File: README.md (hunk 2 above)
Comment (ryanwilsonperkin) [reply README.md:7]: To avoid any potential barrier I'll create all the PRs myself and work with the teams to understand the purpose of them and get them merged. We'll also add the file to wave-cookiecutter-juno so that new apps that are generated will have it by default.

File: index.js (hunk 3 above)
Comment (ryanwilsonperkin) [reply index.js:90]: My mistake, I thought that comment was just being posted in the logs. I can add it in here so that it can add a comment to the PR as well.

**Comment by ryanwilsonperkin** [comment]
Thanks for the review @eballerini! I'll address your comments point by point

> could you summarize the pros and cons of this vs. migbot?

My understanding of GitHub's product structure is that GitHub Actions are the successor to Probot (what migbot is built on) and that Probot can still be useful when you have a specific reason for wanting to maintain a server yourself. When we [last experiences](https://wave.slack.com/archives/CBY8CAVGS/p1584718343001400) an issue with migbot, I was surprised to find that I had to get access to a heroku instance and see logs to understand what was going wrong with it. My personal preference would be to not maintain these servers ourselves, and that's what drew me to GitHub actions as a solution. Here's how I would break down the major pros/cons

Approach | Pros | Cons
--- | --- | --- 
Probot | <ul><li>Very flexible, we control the server so we can do what we like with it.</li><li>Can easily add a database/cache if we require one.</li></ul> | <ul><li>Requires us to maintain a server on Heroku.</li><li>Not clear from looking at the contents of the repo how the Probot app is acting on it (installed as an org-level integration).</li>
Actions | <ul><li>GitHub putting a lot of build/marketing effort behind Action at the moment, lots of new stuff that we can take advantage of.</li><li>Fits the "ephemeral" architectural model of the current feature set of this check.</li><li>Automatically gets a complete copy of the code checked out, can allow us to do more checks with fewer GitHub API calls.</li></ul> | <ul><li>Unfamiliar to most of our developers so far, doesn't have inertia.</li><li>Requires a config file in each repo that wants to take advantage of it (~30)</li></ul>


> what would be the flow to do end-to-end testing for new changes in this repo before pushing them out there? I saw you created the other repo

Creating another repo like I did can be a good approach. We can also install this action on itself and just maintain a `django` and `rails` branch where we perform this type of end-to-end testing locally. Ideally the test suite can also be depended on as we add more edge-case or regression tests to it.

> where is the setting to make this check optional or required? Is it in code?
Same place as it would currently be for the Probot approach. Lives under `https://github.com/waveaccounting/$repo/settings/branch_protection_rules/$branch_protection_id`

> I really like that you added all these unit tests :) this is not a question
Thanks! 😄

**Comment by nickpresta** [comment]
@ryanwilsonperkin do you think this could be a good case for `wave app:scaffold github-action-migration-check`? The generated content is static and has no configuration options, and is easy to use for developers.

**Comment by ryanwilsonperkin** [comment]
@nickpresta yeah this seems like a decent use case for that. It'd pull from the `project_type` to understand if it's a rails or django project, doesn't need much else parameterized. We'd have to talk through the design of an `app:scaffold` command first a bit, like whether it would update things or just create them.

Since this change would be required for people to still get migbot on their repos, I'd still be inclined to do the work for them, but I could script it through `wave app:scaffold` if you think now's a good time to launch that command.

**Comment by nickpresta** [comment]
LMK if you want to chat/go over a document/brainstorm/etc.

File: index.js (hunk 3 above)
Comment (eballerini) [reply index.js:90]: Yes please. That comment is important for people who are not sure why the check failed. Also, we add a label that we rely on to know whether or not a comment has already been added (because we don't want to add a comment for every new commit). It looks like this got lost as well.

**Comment by ryanwilsonperkin** [comment]
@eballerini haha sure thing! How about `migbot-github-action` to clarify that it's a github action that does the "migbot" stuff?

**Comment by ryanwilsonperkin** [comment]
@eballerini alright, should be all up-to-date now with the commenting and labeling behaviour

**Comment by ryanwilsonperkin** [comment]
And name has been updated

**Comment by eballerini** [comment]
> And name has been updated

Would it make sense to rename the test repo as well?

**Comment by ryanwilsonperkin** [comment]
I'll be deleting the test repo after this PR has been merged, and switching over to an approach where we create integration tests within this same repo. We'll create separate distinct branches inside this repo that mimic the content from the test-migration-check-action repo.

File: index.test.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    'project-type': 'django',
+    'GITHUB_TOKEN': 'example',
+  };
+  const inputs = Object.assign({}, DEFAULT_INPUTS, inputOverrides);
+  core.getInput.mockImplementation(key => inputs[key]);
+}
+
+beforeEach(() => jest.resetAllMocks());
+
+test('invalid project fails with error message', async () => {
```
Comment (jcolburne) [inline index.test.js:38]:
```suggestion
test('invalid project-type fails with error message', async () => {
```

**Comment by jcolburne** [comment]
Nice Ryan! One thing I noticed is that we're missing happy path tests for when there is only code added or modified, no migrations.

**Comment by ryanwilsonperkin** [comment]
@jcolburne nice catch! Totally slipped my mind. I've added those test cases now

**Comment by ryanwilsonperkin** [comment]
@eballerini @jcolburne ready for another look when you're ready
